modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/CursorImpl.java

```diff
@@ -159,19 +171,25 @@ public T next() {
             } catch (InterruptedException | ExecutionException e) {
                 Throwable cause = e.getCause();
 
-                if (cause != null) {
-                    if (cause.getClass().equals(NodeStoppingException.class)) {
-                        throw new NoSuchElementException();
-                    } else {
-                        if (cause.getClass().equals(NoSuchElementException.class)) {
-                            throw (NoSuchElementException) cause;
-                        }
-                    }
+                if (cause instanceof NodeStoppingException) {
+                    throw new NoSuchElementException();
+                }
+
+                if (cause instanceof NoSuchElementException) {
+                    throw (NoSuchElementException) cause;
                 }
 
                 LOG.debug("Unable to evaluate cursor hasNext command", e);
 
-                throw new IgniteInternalException(e);
+                if (e.getCause() instanceof IgniteInternalException) {
```

Comment:
I remember you were asking about what to do in such cases. All I can recommend is to have a set of utilitary methods.
I'd go as far as writing it as `throw withCause(MetaStorageException::new, CURSOR_EXECUTION_ERR, e.getCause())`